### ISSUE TITLE
Support encrypted communication with HA

### ISSUE THREAD

**Issue by robbiet480** [home-assistant/home-assistant-android#97]
Just for that extra bit of security, we should implement [encrypted communication with HA](https://developers.home-assistant.io/docs/en/app_integration_sending_data.html#implementing-encryption) over the webhook like iOS supports.

**Comment by robbiet480** [comment]
Some notes on this:

- We need to use an updated and peer reviewed library to do this. @JBassett says he looked at https://github.com/terl/lazysodium-android, one option I suggested. The other option is https://github.com/joshjdevl/libsodium-jni. Sadly, both are Java. The only Kotlin option hasn't been updated in 2+ years.
- Initial support has been added to HA Core to enable encryption on existing registrations (at https://github.com/home-assistant/home-assistant/pull/31743). Call webhook action `enable_encryption` on HA Core 0.106+, receive a JSON body with `secret` as the response. HA version can be derived via `get_config` webhook action. The logic in app should look like this (psuedocode):
    ```
    if app_has_no_encryption_key and ha_version > 0.106:
        attempt_to_enable_encryption()
    ```
- If enabling encryption fails due to older HA version, then we should either keep doing that check every so often (whenever app starts?) or have a button in settings to allow users to manually enable encryption.

**Comment by balloob** [comment]
Do we know the HA version that we talk to ? Seems like something we should know in the app to be able to enable/disable functionality.

**Comment by robbiet480** [comment]
@balloob We can get that info from the `get_config` webhook action.

**Comment by JBassett** [comment]
Initial work https://github.com/home-assistant/home-assistant-android/tree/feature/webhookEncryption

I am running into issues with HA decrypting my requests...  If anyone wants to take a crack at it please do!

**Comment by DirkBaumeister** [comment]
So, I just discovered that the android app in 2026 still does not support encryption of (cloud) webhooks.
Is this feature somewhere on the roadmap?

I think it would be a great addition to have the same security as the iOS App does here.

**Comment by TimoPtr** [comment]
> So, I just discovered that the android app in 2026 still does not support encryption of (cloud) webhooks. Is this feature somewhere on the roadmap?
> 
> I think it would be a great addition to have the same security as the iOS App does here.

The encryption through Cloud is happening using HTTPS, the encryption you are mentioning from iOS is relevant when using HTTP.

**Comment by DirkBaumeister** [comment]
You are mixing two things up here :)

Yes, of course, the transport is encrypted via HTTPS. But you can still send unencrypted payload to a (cloud) webhook over HTTPS created for an android mobile app if you know the right webhook url. When a (cloud) webhook has the "supportEncryption" flag set to true the Home Assistant Core Instance rejects unencrypted payloads and only processes them when the correct secret (Which got generated on register) was provided.

So yes, the connection itself is secure, that's right. But you can still potentially alter with the (cloud) webhooks if you know the correct url. Supporting "supportEncryption" on android (like on iOS) would prevent that.

**Comment by TimoPtr** [comment]
I agree with you that it's 2 different things, but as today the risk is if the URL gets exposed someone else can use it.

**Comment by DirkBaumeister** [comment]
> I agree with you that it's 2 different things, but as today the risk is if the URL gets exposed someone else can use it.

Yes, that's what I meant.
To be fair, a very low risk but still, it would be cool if this potential risk could be eliminated somehow.

**Comment by TimoPtr** [comment]
> > I agree with you that it's 2 different things, but as today the risk is if the URL gets exposed someone else can use it.
> 
> Yes, that's what I meant. To be fair, a very low risk but still, it would be cool if this potential risk could be eliminated somehow.

Indeed, for now it's not a priority for me but we keep this open for when we have a bit of time.

**Comment by DirkBaumeister** [comment]
> > > I agree with you that it's 2 different things, but as today the risk is if the URL gets exposed someone else can use it.
> > 
> > 
> > Yes, that's what I meant. To be fair, a very low risk but still, it would be cool if this potential risk could be eliminated somehow.
> 
> Indeed, for now it's not a priority for me but we keep this open for when we have a bit of time.

Great to here, thanks for your reply! :)
